### PR TITLE
feat: replace range slider with input fields

### LINKED ISSUE #3461: replace range slider with input fields
thanks to @roiLeo and 
- https://github.com/kodadot/nft-gallery/pull/3442

~range slider is passing correct information to query, i.e. priceMin + priceMax works, but we're using implementation from rmrk.
What we need to do now, is adjust the slider to fit BSX needs. RMRK has slider for 0-30 ksm with each tick representing 1ksm increase.~

updated desc: 

- let's throw out the range slider
- let's add input fields for min and max
- make the currency change based on chain
- make sure it looks great on mobile

checkout opensea/looksrare for inspiration

### LINKED ISSUE #3515: toggling 'BUY NOW' should show only prices > 0
@petersopko Oups since my last review, I totaly forgot that not listed nfts have a price of 0, so when we toggle "BUY NOW" it will filter nfts with a price greaterThanOrEqualTo 0 (which corresponds to all nfts)

_Originally posted by @roiLeo in https://github.com/kodadot/nft-gallery/pull/3474#issuecomment-1191383795_

### PR DESCRIPTION
**Pull request by zhengow**
**Thank you for your contribution** to the [KodaDot NFT gallery](https://kodadot.xyz).
👇 \_ Let's make a quick check before the contribution.

### PR type

- [ ] Bugfix
- [x] Feature
- [ ] Refactoring

### What's new?

- [x] PR closes #3461 
- [x] PR closes #3515  
- [ ] Requires deployment <rubick/magick_issue>
- [ ] <brief_description_of_what_I've_added>

### Before submitting Pull Request, please make sure:

- [x] My contribution builds **clean without any errors or warnings**
- [x] I've merged recent default branch -- **main** and I've no conflicts
- [x] I've tried to respect high code quality standards
- [x] I've didn't break any original functionality
- [x] I've posted a screenshot of demonstrated change in this PR

### Optional

- [x] I've tested it at </bsx/explore?page=1&tab=GALLERY&listed=true&min=1&max=100000>
- [ ] I've tested PR on mobile and everything seems works
- [ ] I found edge cases
- [ ] I've written some unit tests 🧪

### Had issue bounty label?

- [x] Fill up your KSM address: [Payout](https://beta.kodadot.xyz/transfer/?target=EgCzZLpVe7zg7RHh79qPyC8UrykVkA22YY2oj7sWMZf15Ws)

### Community participation

- [ ] [Are you at KodaDot Discord?](https://discord.gg/35hzy2dXXh)

### Screenshot

- [x] My fix has changed **something** on UI; a screenshot is best to understand changes for others.
image

image

image

image

image

### CODE REVIEW

**Comment by zhengow** [comment]
> 

Done, I add `type='number'`, it will not allowed to type non-number char.
btw, it will be disabled when `min price > max price`.
I allow empty input, since users maybe need to clearout price range.

File: components/rmrk/Gallery/Search/SearchBar.vue [hunk 1]
```diff
@@ -214,14 +228,39 @@ export default class SearchBar extends mixins(
...
+    const max = this.$route.query.max
+    if (min && max) {
+      return `Prices ranging from ${min} ${this.token} to
+            ${max} ${this.token}`
+    } else if (min && !max) {
+      return `Prices ranging from ${min} ${this.token}`
+    } else if (!min && max) {
+      return `Prices ranging from 0 to ${max} ${this.token}`
+    }
+    return ''
```
[the remark this answers is not in the document]
Comment (roiLeo) [reply components/rmrk/Gallery/Search/SearchBar.vue:261]: can we add translation with values please?

**Review by roiLeo — COMMENTED** [review]
Works really well! 

Here's some comments

would it be possible to format number like we do in price? (10 000)
Capture d’écran 2022-07-19 à 10 21 27 AM

Since you choose to hide inputs when "BUY NOW" is disabled, how users will now they can filter by price range?
Capture d’écran 2022-07-19 à 10 20 25 AM

File: components/rmrk/Gallery/Search/SearchBar.vue (hunk 1 above)
[the remark this answers is not in the document]
Comment (zhengow) [reply components/rmrk/Gallery/Search/SearchBar.vue:261]: > can we add translation with values please?

Do you mean add in `en.json`?

**Comment by zhengow** [comment]
> Since you choose to hide inputs when "BUY NOW" is disabled, how users will now they can filter by price range? Capture d’écran 2022-07-19 à 10 20 25 AM

That's right, I notice that old codes are in this logic. So I need to show them no matter whether it is `buy now`?

**Comment by roiLeo** [comment]
> That's right, I notice that old codes are in this logic. So I need to show them no matter whether it is `buy now`?

This could be done now or latter don't bother with it

**Comment by zhengow** [comment]
> 

I think it still better hide price filter if user choose `no buy now`.
Because some nft has no price if it is not listed.

**Comment by zhengow** [comment]
> > That's right, I notice that old codes are in this logic. So I need to show them no matter whether it is `buy now`?
> 
> This could be done now or latter don't bother with it

Already fix all the advice, pls check again:)

**Review by cryptodamsky — CHANGES_REQUESTED** [review]
The way it’s done on BSX, for example " from 50,000 BSX to 55,000 BSX" is fine, makes sense

But the way it’s done on RMRK is confusing to me.

image

If you have a look, in some countries 1.000 means 1000.

Why have so many 0 there? I would just do " From range 1 to 15"

**Comment by roiLeo** [comment]
@zhengow can you try to enable "BUY NOW" button when we click on "Apply" button?

**Comment by zhengow** [comment]
> @zhengow can you try to enable "BUY NOW" button when we click on "Apply" button? because it's not displaying price on gallery cards

Sure

**Comment by roiLeo** [comment]
> How about raise a new issue for this format problem in `Money Component`?

Depends how you planned to format RMRK values, but we can continue in #3355

**Review by cryptodamsky — COMMENTED** [review]
Feel like it’s not working?

image


or is this the same issue related to the money component? or smth new

**Comment by zhengow** [comment]
> > How about raise a new issue for this format problem in `Money Component`?
> 
> Depends how you planned to format RMRK values, but we can continue in #3355

Fix format RMRK values, looks like this:
image

**Comment by roiLeo** [comment]
Hey don't forget to resolve conflicts please :)

**Comment by roiLeo** [comment]
> > Still same issue?
> 
> Screenshot 2022-07-20 at 15-55-49 Low minting fees and carbonless NFTs KodaDot - Kusama NFT Market Explorer
> 
> Same for me, please don't use tooltip to display error message, [Buefy](https://buefy.org/documentation/input/#validation) inputs has validation or you can provide messages to `<b-field />` (plus there is a typo in your text)

**Comment by zhengow** [comment]
> Still same issue?
> 
> image

My bugs, for not change type of variables `min` and `max` and compare them. Now fix it:

image
image

**Comment by zhengow** [comment]
> > Still same issue?
> 
> Screenshot 2022-07-20 at 15-55-49 Low minting fees and carbonless NFTs KodaDot - Kusama NFT Market Explorer
> 
> Same for me, please don't use tooltip to display error message, [Buefy](https://buefy.org/documentation/input/#validation) inputs has validation or you can provide custom messages to `<b-field />` (plus there is a typo in your text)

Show validation massage in the bottom:

image

File: components/rmrk/Gallery/Gallery.vue [hunk 2]
```diff
@@ -358,29 +358,32 @@ export default class Gallery extends mixins(
...
-        params.push({
-          price: {
-            greaterThan: this.searchQuery.priceMin,
-            lessThanOrEqualTo: this.searchQuery.priceMax,
-          },
-        })
+        if (this.searchQuery.priceMax) {
+          params.push({
+            price: {
+              greaterThan: minPrice,
```
Comment (roiLeo) [inline components/rmrk/Gallery/Gallery.vue:367]:
```suggestion
              greaterThanOrEqualTo: minPrice,
```

File: components/rmrk/Gallery/Gallery.vue [hunk 3]
```diff
@@ -358,29 +358,32 @@ export default class Gallery extends mixins(
...
+              lessThanOrEqualTo: this.searchQuery.priceMax,
+            },
+          })
+        } else {
+          params.push({
+            price: { greaterThan: minPrice },
```
Comment (roiLeo) [inline components/rmrk/Gallery/Gallery.vue:373]:
```suggestion
            price: { greaterThanOrEqualTo: minPrice },
```

File: components/rmrk/Gallery/Gallery.vue [hunk 4]
```diff
@@ -358,29 +358,32 @@ export default class Gallery extends mixins(
...
+          })
+        }
       } else {
-        params.push({
-          price_gt: this.searchQuery.priceMin,
-          price_lte: this.searchQuery.priceMax,
-        })
+        if (this.searchQuery.priceMax) {
+          params.push({
+            price_gt: minPrice,
```
Comment (roiLeo) [inline components/rmrk/Gallery/Gallery.vue:379]:
```suggestion
            price_gte: minPrice,
```

File: components/rmrk/Gallery/Gallery.vue [hunk 5]
```diff
@@ -358,29 +358,32 @@ export default class Gallery extends mixins(
...
+            price_lte: this.searchQuery.priceMax,
+          })
+        } else {
+          params.push({
+            price_gt: minPrice,
```
Comment (roiLeo) [inline components/rmrk/Gallery/Gallery.vue:384]:
```suggestion
            price_gte: minPrice,
```

File: components/rmrk/Gallery/Search/SearchBar.vue [hunk 6]
```diff
@@ -128,20 +126,37 @@
...
-        :min="0"
-        :step="1"
-        ticks
-        @change="sliderChange">
-      </b-slider>
+      <div v-if="!hideFilter">
+        <b-field class="columns mb-0">
+          <b-input
+            type="number"
+            min="0"
```
Comment (roiLeo) [inline components/rmrk/Gallery/Search/SearchBar.vue:133]:
```suggestion
            min="0"
            step="any"
```

File: components/rmrk/Gallery/Search/SearchBar.vue [hunk 7]
```diff
@@ -128,20 +126,37 @@
...
+            class="column is-2"
+            :placeholder="$t('query.priceRange.minPrice')"
+            v-model="rangeSlider[0]">
+          </b-input>
+          <b-input
+            min="0"
```
Comment (roiLeo) [inline components/rmrk/Gallery/Search/SearchBar.vue:139]:
```suggestion
            min="0"
            step="any"
```

**Review by roiLeo — CHANGES_REQUESTED** [review]
Hey! Tested on bsx & rmrk

Small last changes, see review
Screenshot 2022-07-21 at 10-56-52 Low minting fees and carbonless NFTs KodaDot - Kusama NFT Market Explorer

 - It throw errors when i type decimal number. Fix can be using `step` attribute (ref [spec](https://html.spec.whatwg.org/multipage/forms.html#the-step-attribute))
 - when filter minimum 20bsx, I would like to see items >= 20bsx

**Comment by zhengow** [comment]
> Hey! Tested on bsx & rmrk
> 
> Small last changes, see review Screenshot 2022-07-21 at 10-56-52 Low minting fees and carbonless NFTs KodaDot - Kusama NFT Market Explorer
> 
> * It throw errors when i type decimal number. Fix can be using `step` attribute (ref [spec](https://html.spec.whatwg.org/multipage/forms.html#the-step-attribute))
> * when filter minimum 20bsx, I would like to see items >= 20bsx

Fix it:)

**Review by roiLeo — APPROVED** [review]
✅ Tested & code lgtm

Thank you for your good work!

**Review by petersopko — APPROVED** [review]
wfm

**Comment by roiLeo** [comment]
Oups since my last review, I totaly forgot that not listed nfts have a price of 0, so when we toggle "BUY NOW" it will filter nfts with a price greaterThanOrEqualTo 0 (which corresponds to all nfts)

**Comment by petersopko** [comment]
> @petersopko Oups since my last review, I totaly forgot that not listed nfts have a price of 0, so when we toggle "BUY NOW" it will filter nfts with a price greaterThanOrEqualTo 0 (which corresponds to all nfts)

let's deal with it in 
- #3515

**Comment by roiLeo** [comment]
wait i'm currently on this branch

**Comment by petersopko** [comment]
> wait i'm currently on this branch

ah great, I'll wait for this one then :)

**Comment by roiLeo** [comment]
noooo :(

**Comment by roiLeo** [comment]
> noooo :(

now fixed! sorry for that but everytime someone enable merge from main I can't push anymore (not a fan of rebase)

**Comment by petersopko** [comment]
> now fixed! sorry for that but everytime someone enable merge from main I can't push anymore (not a fan of rebase)

I didn't see your comment before I updated the branch, sorry 😆 

btw it seems to have issues even in the other way around? when I untoggle 'BUY NOW' the filter for gt.than > 0 disappears, but when I'm sorting by price desc.; I get to see NFTs with highest price (but price isn't showing in the gallery card).

Maybe this is something we should fix elsewhere, if the price is 0 we don't need the gallery card desc. but if it's more than that, we should show the price regardless of the 'BUY NOW' toggle position

current state:
image

desired outcome:
image

**Comment by roiLeo** [comment]
> btw it seems to have issues even in the other way around? when I untoggle 'BUY NOW' the filter for gt.than > 0 disappears, but when I'm sorting by price desc.; I get to see NFTs with highest price (but price isn't showing in the gallery card).

Always been like this, no? 
https://kodadot.xyz/rmrk/explore?page=1&tab=GALLERY&listed=false&sort=PRICE_DESC

> Maybe this is something we should fix elsewhere, if the price is 0 we don't need the gallery card desc. but if it's more than that, we should show the price regardless of the 'BUY NOW' toggle position

Or we could try to toggle "BUY NOW" when sorting by PRICE and it would show the price on cards

**Comment by petersopko** [comment]
pay 200 usd

**Comment by yangwao** [comment]
😍 Perfect, I’ve sent the payout 
💵 $200 @ 59.07 USD/KSM ~ 3.386 $KSM
🧗 [EgCzZLpVe7zg7RHh79qPyC8UrykVkA22YY2oj7sWMZf15Ws](https://kusama.subscan.io/account/EgCzZLpVe7zg7RHh79qPyC8UrykVkA22YY2oj7sWMZf15Ws)
🔗 [0x9eb083b9206254a74695485cdfa3c312eaf6d7cfdebd19602c988df2d747a364](https://kusama.subscan.io/extrinsic/0x9eb083b9206254a74695485cdfa3c312eaf6d7cfdebd19602c988df2d747a364)

🪅 Let’s grab another issue and get rewarded! 
🪄 [github.com/kodadot/nft-gallery/issues](https://github.com/kodadot/nft-gallery/issues/)

**Pull request merged**